File: lpfgopt/src/lpfgopt.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>
#include "lpfgopt.h"


//typedef struct two_dimensional_array{
/* Structure for 2d arrays to be used in the optimization */
//    int rows;
//    int columns;
//   double** array;
//}array_2d;


void LPFGOPTCALL free_array_2d(array_2d* array)
{
/* Function for freeing mallocs for two_dimensional_array */
    for (int i = 0; i < array->rows; i++){
        free(array->array[i]);
        }
    free(array->array);
    free(array);
    /*Good rule to remember: For every 
      malloc calloc or realloc you must have one free*/
}

double uniform(double lower, double upper)
{
/* Generate a random number bewtween lower and upper */
    int rint;
    double zero_one, rinp;
    rint = rand();
    zero_one = 1.0 * rint / RAND_MAX;
    rinp = (upper - lower) * zero_one + lower;
    
    return rinp;
}


array_2d* LPFGOPTCALL zeros(int rows, int columns)
{
/* Generate a 2d array with rows rows and columns columns */
    int i;
    int j;
    array_2d* array1 = (array_2d*) malloc(sizeof(array_2d));
    double** the_array = (double**) malloc(rows*sizeof(double*));

    for (i = 0; i < rows; i++){
        the_array[i] = (double*) malloc(columns*sizeof(double));
        for (j = 0; j < columns; j++){
            the_array[i][j] = 0.0;
            }
        }

    array1->array   = the_array;
    array1->rows    = rows;
    array1->columns = columns;
    
    return array1;
}


void gen_pt_set(array_2d* matrix, array_2d* intervals,
                double (*func)(double*), double* args)
{
/* Poplulate a matrix with column 0 being the values of the 
   objective function and the other columns being the arguments
   of that function*/
    int row;
    int col;
    double rinp;   
    for (row = 0; row < matrix->rows; row++){
        for (col = 1; col < matrix->columns; col++){
            rinp = uniform(intervals->array[col-1][0],
                intervals->array[col-1][1]);
            matrix->array[row][col] = rinp;
            args[col-1] = rinp;
            }
        //rinp = func(args); //##ISSUE!!!
        matrix->array[row][0] = func(args); 
        }

}

void eval_bw(array_2d* matrix, int* bw_index)
{
/* Get the indices of the best and worst objective values 
   of the objective function */
    int b = 0;
    int w = 0;
    int i;

    for (i = 0; i < matrix->rows; i++){
        if (matrix->array[i][0] < matrix->array[b][0]) b = i;
        if (matrix->array[i][0] > matrix->array[w][0]) w = i;
        }
    bw_index[0] = b;
    bw_index[1] = w;
}


void leap_frog(double (*func)(double*), array_2d* matrix,
               int bi, int wi, array_2d* int_1, double* args)
{
/* Run one iteration of the Leapfrog algorithm given the matrix
   the objective function, intervals and the indices of the best
   and worst objective function values */
    int i;
    double itvl1;
    double itvl2;
    double lower;
    for (i = 0; i < (matrix->columns - 1); i++){
        // For each argument, choose a new interval over which
        // to leap frog based on the opposite of the difference
        // between the best and worst points
        itvl1 = matrix->array[bi][i+1];
        itvl2 = 2.0 * matrix->array[bi][i+1] - matrix->array[wi][i+1];
        if (itvl1 > itvl2){
            // Sort the two points so the lower is itvl1
            lower = itvl2;
            itvl2 = itvl1;
            itvl1 = lower;
            }
        // Check and correct for intervals going out of bounds
        if (itvl1 < int_1->array[i][0]) itvl1 = int_1->array[i][0];
        if (itvl2 > int_1->array[i][1]) itvl2 = int_1->array[i][1];
        args[i] = uniform(itvl1, itvl2); // Get a random input
        matrix->array[wi][i+1] = args[i];// Update matrix and args
        }
    matrix->array[wi][0] = func(args); // Update objective ISSUE!!!
}


double err_calc(array_2d* matrix, int bi, int wi)
{
    int i;
    int j;
    double err;

    err = fabs(matrix->array[bi][0] - matrix->array[wi][0]);

    for (i = 0; i < matrix->rows; i++){
        for (j = 1; j < matrix->columns; j++){
            err += fabs(matrix->array[bi][j] - matrix->array[i][j]);
            }
        }
        
    return err;
    /* relative error returned overflow errors */
}


double* LPFGOPTCALL minimize(double (*f1_ptr)(double*), array_2d* int_1,
              int npts, int nvars)
{
/* Minimize function that runs the leap_frog algorithm to
   minimize f1_ptr on the intervals on int_1 */
    int bi;
    int wi;
    int iter;
    int dvi;
    double err;
    // initalize maximum iterations and the tolerance;
    int maxit = 10000;
    double tol = 1E-5;
    // Heap allocate our pt_set our best and worst indexes and 
    // our arg array;
    array_2d* pt_set;
    int* bw_index = (int*)malloc(2*sizeof(int));
    double* args = (double*) malloc(nvars * sizeof(double));
    
    
    srand(time(NULL)); // Seed random number generator;
    // Generate an empty matrix and populate it with
    // objective values and random points on the intervals;
    pt_set = zeros(npts,nvars+1);
    gen_pt_set(pt_set, int_1, f1_ptr, args);
    
    // Iterate to minimize
    for (iter = 0; iter < maxit; iter++){
        eval_bw(pt_set, bw_index); // Get best and worst points;
        bi = bw_index[0];
        wi = bw_index[1];
        // Leapfrog the worst over the best
        leap_frog(f1_ptr, pt_set, bi, wi, int_1, args);
        err = err_calc(pt_set, bi, wi); // Calculate error
        
        if (err <= tol) break; // Check for convergence
        }

    if (iter == maxit){ 
        printf("Warning: maximum iterations exceeeded.\n");
        } // If maxit is reached warn user 
    
    
    double* best = (double*) malloc(nvars * sizeof(double));
    
    for (dvi = 0; dvi < nvars + 1; dvi++){
        best[dvi] = pt_set->array[bi][dvi];
        }
    // Free heap allocations
    
    free(bw_index);
    free(args);
    free_array_2d(pt_set);
    //printf("RUN SUCCESSFUL!!\n");
    return best;
   
}


/*
int main(int argc, char const* argv[])
{
    //clock_t begin = clock();
    
    array_2d* int_1;
    int nx = 20;
    int ny = 3;
    double (*f1_ptr)(double*);
    f1_ptr = &f1;

    int_1 = zeros(ny,2);
    int_1->array[0][0] = -10.0;
    int_1->array[0][1] =  10.0;
    int_1->array[1][0] = -10.0;
    int_1->array[1][1] =  10.0;
    
    
    minimize(f1_ptr, int_1, nx, ny);
    free_array_2d(int_1);
    
    //clock_t end = clock();
    //double time_spent = (double)(end - begin) / CLOCKS_PER_SEC;
    //printf("Optimization performed in %.9f seconds.\n", time_spent);
    // Optimization performed in 0.003000000 seconds. (Windows)
    // This is 7.3 times faster than python did the optimization (Linux: 0.022 sec)
    // on linux it was clocked at .002 seconds which makes it 11 times faster
    // than python
    
    return 0;
}*/

File: lpfgopt/lpfgopt.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

"""
file name      : lpfgopt.py
title          : Leap Frog Optimizer
version        : 0.6.2 ALPHA
last modified  : 11 April 2017 
author         : Mark Redd
email          : redddogjr@gmail.com

written for python version: 2

optimizer algorithm website: http://www.r3eda.com/

about:

    This optimizer was written based on the algorithm published by
    Dr. R. Russell Rhinehart.

    A full explanation of the algorithm can be found at the following URL:

    http://www.r3eda.com/leapfrogging-optimization-algorithm/

    The following are "key references" published on
    the optimization website explaining the technique:

      - Rhinehart, R. R., M. Su, and U. Manimegalai-Sridhar,
        “Leapfrogging and Synoptic Leapfrogging: a new optimization approach”,
        Computers & Chemical Engineering, Vol. 40, 11 May 2012, pp. 67-81.

      - Manimegalai-Sridhar, U., A. Govindarajan, and R. R. Rhinehart,
        “Improved Initialization of Players in Leapfrogging Optimization”,
        Computers & Chemical Engineering, Vol. 60, 2014, 426-429.

      - Rhinehart, R. R.,
        “Convergence Criterion in Optimilsation of Stochastic Processes”,
        Computers & Chemical Engineering, Vol. 68, 4 Sept 2014, pp 1-6.

How to use the optimizer:
    Needed from user: Intervals for all variables, the function that needs to be
    minimized, other constraints
"""

from warnings import warn
from numpy.random import uniform
from numpy import prod, abs as npabs, sum as npsum, argsort, zeros, array

def _err_calc(b, w, pt_set, rel=True):
    """
    __PRIVATE FUNCTION__
    Calculate the error of the system to compare against the tolerance for
    convergence. In this scope error is defined as the sum of:
    
     - the absolute or relative error of the objective function between the best 
        and worst points
     - the sum of the absolute difference between each decision variable and 
        the corresponding variable value of the best point for every point.
    
    or mathmatically:
    
    err = |(f(w) - f(b))/f(b)| + sum_i=0(sum_j=0(|(b_ij-p_ij)/b_ij|)
    
    for i points and j variables.    
    """
    if rel:                # select relative or absolute error
        if prod(b) == 0.0: # avoid divide by zero errors
            b += 1.0e-8
        norm1 = b[0]
        norm2 = b[:1]
    else:
        norm1 = 1.0
        norm2 = 1.0

    err_obj = npabs((w[0] - b[0])/norm1)            # Get the objective error
    dist_sum = 0
    for i in pt_set:                                 # sum over points
        dist = npsum(npabs((b[1:] - i[1:])/norm2)) # sum over variables
        dist_sum += dist

    return err_obj + dist_sum


def _eval_bw(pt_set):
    """
    __PRIVATE FUNCTION__
    Return the indices of the best and worst points in the point set(pt_set).
    Method: Starts by generating a sorted index array based on column 0
    and then returns the first and last elements.
    """
    index1 = argsort(pt_set[:,0])
    return index1[0], index1[-1]


def _gen_pt_set(f, intervals, points=20, constraints=None, args=()):
    """
    __PRIVATE FUNCTION__
    Generates a point set with "points" rows and "len(intervals)+1" columns
    This matrix will be the initial point set that will be used in the Leapfrog
    Algorithm. 
    Args:
        f           - The objective function to be optimized
        intervals   - A 2d array with n rows and 2 columns with n being the 
                      number of decision variables and each column being the 
                      lower and upper bounds on each variable respecively.
        points      - The number of randomly generated points to be used in the
                      optimization.
        constraints - NOT YET FUNCTIONAL
    
    The output matrix will have columns of the following form:
    
        [f(x), x[0], x[1], x[2],...x[n-1]]
    """
    number_of_cols = len(intervals) + 1
    pt_set = zeros([points, number_of_cols])
    for n in xrange(points):                # for each point make a vector
        for v in xrange(1,number_of_cols):  # for each variable
                                            
            pt_set[n][v] = uniform(         # get a random input in the interval
                intervals[v-1][0], 
                intervals[v-1][1])
            pt_set[n][0] = f(pt_set[n][1:], *args) # get objective value
    return pt_set


def _get_new_w(f, b, w, intvl, args=()):
    """
    __PRIVATE FUNCTION__
    Picks a new random interval by "Leapfrogging" the worst point over the best
    point. The new point is randomly chosen in the domain opposite the best 
    point with limits being the difference between the worst and best points and
    the bounds set by intervals in intvl. 
    """
    n_vars = len(w)-1           ## number of variables
    inputs = zeros(n_vars)   ## make an input array
    for i in xrange(n_vars):
        ## pick an interval for w that is opposite b
        new_interval = array([b[i+1] ,2*b[i+1] - w[i+1]])
        new_interval.sort() ## put the lower limit first
        ## checking that interval does not go out of bounds
        if new_interval[0] < intvl[i][0]:
            new_interval[0] = intvl[i][0] ## if the interval is out of bounds    
        if new_interval[1] > intvl[i][1]: # change the upper or lower limit
            new_interval[1] = intvl[i][1] # to be the edge of the bounds
        w[i+1] = uniform(new_interval[0], new_interval[1]) ## get a new input
        inputs[i] = w[i+1]
    w[0] = f(inputs, *args) ## get a new objective value from the new inputs
    return w


def minimize(f,intervals,args=(),full_output=False, points=20,
    constraints=None, tol=1e-5, rel=True, maxit=10000):
    """
    Minimize f on the array of intervals for each decision variable.
    
    Args:
        f           - The objective function to be minimized
        args        - Other arguments to pass into objective function
        full_output - When set to true the function returns a dictionary of the
                        pertinent data from the optimization including:
                        
                        * best       : best point in the point set
                        * worst      : worst point in the point set
                        * final_error: final overall error
                        * iterations : iterations to convergence or max
                        * point_set  : the final state of the point set
                        
        points      - Number of points to be used in the optimization. Default
                        is 20 points.
        constraints - NOT YET FUNCTIONAL
        tol         - Convergence tolerance or maximum error to converge.
        maxit       - Maximum iterations before returning. If maxit is reached
                        the system returns a runtime warning.
    """
    pt_set = _gen_pt_set(f,intervals,points, args=args) # generate the point set

    for iters in xrange(maxit):

        bi, wi = _eval_bw(pt_set)  # Get the indices of the best and worst points
        b = pt_set[bi]
        w = pt_set[wi]
        w_new = _get_new_w(f,b,w,intervals, args=args) 
        pt_set[wi] = w_new                # Replace the worst with the new point
        err = _err_calc(b, w_new, pt_set, rel) # Get the overall error
        
        if err < tol:                     # Check for convergence
            if full_output == True:
                return {
                    "best":         b,
                    "worst":        w,
                    "final_error":  err,
                    "iterations":   iters + 1,
                    "point_set":    pt_set
                    }
            else:
                return b

    warn("Optimizer: Maximum Iterations Exceeded", RuntimeWarning)
    
    if full_output == True:
        return {
            "best":         b,
            "worst":        w,
            "final_error":  err,
            "iterations":   iters + 1,
            "point_set":    pt_set
            }
    else:
        return b

def lfsolve(f,x0=None,intervals=None,args=(),full_output=False,
        points=20, constraints=None, tol=1e-5, s_step=None, maxit=10000):
    """
    A numerical solver based on the Leapfrogging algorithm.
    Solver works by minimizing the sum of the absolute value of the outputs.
    This should return an optimal value of zero if the root is bracketed.
    
    Inputs may be intervals that bracket the root or a guess value vector from 
    which intervals are guessed and then made larger if a root is not found. 
    This is done by calculating a spread step and then expanding as needed 
    until a root is found.
    
    Args:
        f           - Solver function f(x) = 0
        args        - Other arguments to pass into objective function
        full_output - When set to true the function returns a dictionary of the
                        pertinent data from the optimization including:
                        
                        * root        : Values of the vector x such that:
                                        f(x) = 0
                        * final_error : Final value of f_obj
                        * opt_iters   : Final number of iterations in the 
                                        optimizer to find the root
                        * sol_iters   : Number of iterations in the solver 
                                        needed to bracket the root
                        * point_set   : Final point set from the optimizer
                        
        points      - Number of points to be used in the optimization. Default
                        is 20 points.
        constraints - NOT YET FUNCTIONAL
        tol         - Convergence tolerance or maximum error to converge.
        maxit       - Maximum attempts to bracket the root before returning. 
                        If maxit is reached the system returns a 
                        runtime warning.
    """
    # define the solver objective function
    def f_obj(x,*args): return npsum(npabs(f(x,*args))) 
    
    it1 = 0 # initialize the solver iterations
    
    if x0 == None and intervals == None: 
        # check if nothing is specified for guess or intervals
        raise ValueError("Initial guess (x0) or intervals must"\
            " be specified.")

    elif intervals == None: # if a guess value for each DV is specified
        nvs = len(x0)       # number of variables = length of the guess vector
        
        if s_step == None:  # if no spread step is specified estimate one based
            s_step = f_obj(x0,*args) * 0.1 # on the current value of the 
                                     # objective function
        spread = 0.0
        
        for it1 in xrange(maxit):
            spread += s_step                # guess an interval centered on x0
            intervals = zeros([nvs,2])   # with bounds = x0 +- spread
            for i in xrange(nvs):
                intervals[i,0] = x0[i] - spread
                intervals[i,1] = x0[i] + spread
            
            # run the optimization
            root = minimize(f_obj,intervals=intervals,args=args,
                full_output=True, points=points, constraints=constraints,
                tol=tol,rel=False, maxit=maxit)

            root_err = root['best'][0] # get the distance from zero
            
            if root_err < tol:         # check for convergence
                if full_output == True:
                    return {
                        "root":         root['best'][1:],
                        "final_error":  root['best'][0],
                        "opt_iters":    root['iterations'],
                        "sol_iters":    it1 + 1,
                        "point_set":    root["point_set"]
                        }
                else:
                    return root['best'][1:]
                    
        warn("Solver: Maximum Iterations Exceeded. \nSolver was unable to"\
             "bracket the root.", RuntimeWarning)

    else:
        # run the optimization based on the intervals specified
        root = minimize(f_obj,intervals=intervals,args=args,full_output=True,
                points=points, constraints=constraints, tol=tol,
                rel=False, maxit=maxit)
    
    if full_output == True:
        return {
            "root":         root['best'][1:],
            "final_error":  root['best'][0],
            "opt_iters":    root['iterations'],
            "sol_iters":    it1 + 1,
            "point_set":    root["point_set"]
            }
    else:
        return root['best'][1:]

#    def int_minimize(f,x0=None,intervals=None,args=(),full_output=False,
#        points=20, constraints=None, tol=1e-5, s_step=None, maxit=10000):
#        
#        pass

def _main():
    """__PRIVATE FUNCTION__ Some driver code to test the algorithm."""
    
    def test(x)      : return x[0]**2.0 + x[1]**2.0 + 3.0
    #def solve_test(x): return array([2 + x[0], 3 - x[1]])
    #def solve_test2(x, p): return array([2 + x[0], 3 - x[1] + p])
    
    int1 = [
        [-10.0,10.0],
        [-10.0,10.0]]
    x = minimize(test, int1, full_output=True)
    print '\nbest:', x['best'], '\n'
"""    
    x0 = [-3.0,3.0]
    x = lfsolve(solve_test2, x0, args=(3.0,), full_output=True)
    print x['opt_iters']
    print x['sol_iters']
    print 'root:', x['root'], '\n'

"""
"""
real    0m1.127s
user    0m1.056s
sys     0m0.022s
"""

if __name__ == "__main__":
    _main()

